Handle errors on finalizing draft order

orzechdev opened the pull request:
It adds errors handling in the draft order view.

<!-- Please mention all relevant issue numbers. -->

**PR intended to be tested with API branch:** <!-- For example: feature/warehouses  --> main

### Screenshots

<!-- If your changes affect the UI, providing "before" and "after" screenshots will
greatly reduce the amount of work needed to review your work. -->
<img width="1259" alt="Screenshot 2022-07-25 at 16 19 50" src="https://user-images.githubusercontent.com/9825562/180807939-2d78e4cd-89c5-4bab-9c76-0e351edfbedb.png">

### Pull Request Checklist

<!-- Please keep this section. It will make maintainer's life easier. -->

1. [ ] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

<!-- Do not remove this section. It is required to properly setup test deployment instance.
Modify API_URI if you want test instance to use custom backend. CYPRESS_API_URI is optional, use when necessary. -->

API_URI=https://automation-dashboard.staging.saleor.cloud/graphql/
MARKETPLACE_URL=https://marketplace-gray.vercel.app/

patrys commented:
Task linked: [SALEOR-1879 Handle Errors on finalizing to Draft Order](https://app.clickup.com/t/2549495/SALEOR-1879)

Cloud11PL commented on src/orders/components/OrderDraftDetailsProducts/OrderDraftDetailsProducts.tsx:
```diff
@@ -20,60 +27,65 @@ export interface FormData {
+  const intl = useIntl();
   const classes = useStyles(props);
 
+  const formErrors = errors.filter(error => error.field === "lines"); // getFormErrors(["lines"], errors);
```
The comment is not necessary I think

Cloud11PL commented on src/orders/components/OrderDraftDetailsProducts/TableLineAlert.tsx:
```diff
@@ -0,0 +1,37 @@
+    ) : (
+      <ul>
+        {alerts.map((alert, index) => (
+          <li key={index}>{alert}</li>
```
Maybe like this?
```suggestion
          <li key={`${index}_${alert}`}>{alert}</li>
```

Cloud11PL commented on src/orders/components/OrderDraftDetailsProducts/messages.ts:
```diff
@@ -0,0 +1,14 @@
+    defaultMessage: "This product is not published in this channel.",
+    description: "alert message",
+  },
+  notAvailable: {
+    id: "zO+l0L",
+    defaultMessage: "This product is not available for sale in this channel.",
+    description: "alert message",
+  },
```
The same message is in `locale/defaultMessages.json`

orzechdev replied:
`locale/defaultMessages.json` are auto-generated on purpose for service like Transifex from all ts/tsx files.

Cloud11PL replied:
Ah, alright!

Cloud11PL commented on src/orders/components/OrderDraftPage/OrderDraftAlert.tsx:
```diff
@@ -0,0 +1,71 @@
+  order?: OrderDetailsFragment,
+  channelUsabilityData?: ChannelUsabilityDataQuery,
+) => {
+  const inactiveChannel = order && !order.channel.isActive;
```
```suggestion
  const isChannelInactive = order && !order.channel.isActive;
```
or just `isChannelActive`

Cloud11PL commented on src/orders/components/OrderDraftPage/OrderDraftAlert.tsx:
```diff
@@ -0,0 +1,71 @@
+
+  return (
+    <Alert variant="warning" close className={classes.root} {...alertProps}>
+      {alerts.length > 1 ? (
+        <>
+          <FormattedMessage {...alertMessages.manyAlerts} />
+          <ul>
+            {alerts.map((alert, index) => (
+              <li key={index}>
+                <FormattedMessage {...alert} />
+              </li>
+            ))}
+          </ul>
+        </>
+      ) : (
+        <FormattedMessage {...alerts[0]} />
+      )}
```
It's very similar logic to what's in `src/orders/components/OrderDraftDetailsProducts/TableLineAlert.tsx`. Maybe we can unify it somehow (if it makes sense ofc)

orzechdev replied:
Ok, I refactored it a bit

Cloud11PL replied:
Looks good 👍

dominik-zeglen commented on src/orders/components/OrderCustomer/AddressFields.tsx:
```diff
@@ -0,0 +1,31 @@
+import { Typography } from "@material-ui/core";
```
I thought we have address form component somewhere

dominik-zeglen commented on src/orders/components/OrderCustomer/OrderCustomer.tsx:
```diff
@@ -121,6 +110,15 @@ const OrderCustomer: React.FC<OrderCustomerProps> = props => {
     return "";
   };
 
+  const addressTextError = (orderError: OrderErrorFragment) => (
+    <>
+      <Typography variant="body2" className={classes.textError}>
+        {getOrderErrorMessage(orderError, intl)}
+      </Typography>
+      <FormSpacer />
+    </>
+  );
```
Either make it component or copy it, component factories are antipattern in react

dominik-zeglen commented on src/orders/components/OrderDraftDetailsProducts/useLineAlerts.ts:
```diff
@@ -0,0 +1,41 @@
+    alerts.push(intl.formatMessage(lineAlertMessages.notAvailable));
+  }
+
+  return alerts;
```
If you're making it a hook then please at least memoize its output

timuric commented:
Amazing PR!

Some edge cases:

I get this message 
<img width="1335" alt="image" src="https://user-images.githubusercontent.com/4006792/182387214-e52cfbe8-7632-4311-aeba-5b54e64dc2ce.png">

but then it disappears when I add customer address
<img width="1320" alt="image" src="https://user-images.githubusercontent.com/4006792/182386994-9a194b31-1470-4705-86f3-10ae5a48f193.png">

timuric commented:
Maybe we should replace error message with something more friendly. 

`No products are available in channel of this order`
`Channel_name`

<img width="1852" alt="image" src="https://user-images.githubusercontent.com/4006792/182387691-0416cc6e-66ad-45c1-bf49-bda810ced97d.png">

orzechdev commented:
@timuric 
Ad.1. Changed. The warning is now only returned when shipping address and at least one line is set, as backend cannot determine what shipping methods are available without these data. This is why the warning could be wrongly displayed.
Ad.2. Changed appropriately:
<img width="624" alt="Screenshot 2022-08-04 at 14 04 32" src="https://user-images.githubusercontent.com/9825562/182845545-f6686680-2a93-41ea-9383-c4afdeb9b4ea.png">
<img width="627" alt="Screenshot 2022-08-04 at 14 04 46" src="https://user-images.githubusercontent.com/9825562/182845562-a18eab52-454e-49f3-b623-e7cdc1525554.png">

patrys commented:
Could we explicitly state the channel when we're signaling a potentially broken setup? If my colleague sent me this screenshot, there is no way for me to guess what needs to be fixed.

orzechdev commented:
I've updated the messages:
![Screenshot 2022-08-05 at 14 26 51](https://user-images.githubusercontent.com/9825562/183079603-56a40fc1-95f8-43c2-a10d-310a08238934.png)
![Screenshot 2022-08-05 at 14 27 15](https://user-images.githubusercontent.com/9825562/183079612-26964d13-3af5-4a9f-9b55-06e6be262dc5.png)

Also I've also changed channel name from text to link and move this section to the top ("Channel without products" is a channel name here):
![Screenshot 2022-08-05 at 14 45 40](https://user-images.githubusercontent.com/9825562/183080097-05170e81-c011-4889-8f32-6c67b950bb0b.png)

I would stay with messages "... channel assigned to this order" instead of explicitly naming channel to indicate what this channel has to do with this order.

Is it better right now @patrys ?

patrys commented:
I think it's good.
